// @shopify/hydrogen-codegen/dist/cjs/index.cjs
'use strict';

var preset_js = require('./preset.cjs');
var plugin_js = require('./plugin.cjs');
var schema_js = require('./schema.cjs');
var sources_js = require('./sources.cjs');
var pluck_js = require('./pluck.cjs');



Object.defineProperty(exports, 'preset', {
  enumerable: true,
  get: function () { return preset_js.preset; }
});
Object.defineProperty(exports, 'plugin', {
  enumerable: true,
  get: function () { return plugin_js.plugin; }
});
Object.defineProperty(exports, 'getSchema', {
  enumerable: true,
  get: function () { return schema_js.getSchema; }
});
Object.defineProperty(exports, 'schema', {
  enumerable: true,
  get: function () { return schema_js.schema; }
});
Object.defineProperty(exports, 'processSources', {
  enumerable: true,
  get: function () { return sources_js.processSources; }
});
Object.defineProperty(exports, 'patchGqlPluck', {
  enumerable: true,
  get: function () { return pluck_js.patchGqlPluck; }
});
Object.defineProperty(exports, 'pluckConfig', {
  enumerable: true,
  get: function () { return pluck_js.pluckConfig; }
});
