// @shopify/hydrogen-codegen/dist/cjs/preset.cjs
'use strict';

var addPlugin = require('@graphql-codegen/add');
var typescriptOperationPlugin = require('@graphql-codegen/typescript-operations');
var sources_js = require('./sources.cjs');
var plugin_js = require('./plugin.cjs');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var addPlugin__namespace = /*#__PURE__*/_interopNamespaceDefault(addPlugin);
var typescriptOperationPlugin__namespace = /*#__PURE__*/_interopNamespaceDefault(typescriptOperationPlugin);

const namespacedImportName = "StorefrontAPI";
const interfaceExtensionCode = `
declare module '@shopify/hydrogen' {
  interface StorefrontQueries extends GeneratedQueryTypes {}
  interface StorefrontMutations extends GeneratedMutationTypes {}
}`;
const preset = {
  buildGeneratesSection: (options) => {
    if (!options.baseOutputDir.endsWith(".d.ts")) {
      throw new Error("[hydrogen-preset] target output should be a .d.ts file");
    }
    if (options.plugins?.length > 0 && Object.keys(options.plugins).some((p) => p.startsWith("typescript"))) {
      throw new Error(
        "[hydrogen-preset] providing additional typescript-based `plugins` leads to duplicated generated types"
      );
    }
    const sourcesWithOperations = sources_js.processSources(options.documents);
    const sources = sourcesWithOperations.map(({ source }) => source);
    const pluginMap = {
      ...options.pluginMap,
      [`add`]: addPlugin__namespace,
      [`typescript-operations`]: typescriptOperationPlugin__namespace,
      [`gen-dts`]: { plugin: plugin_js.plugin }
    };
    const plugins = [
      {
        [`add`]: {
          content: `/* eslint-disable eslint-comments/disable-enable-pair */
/* eslint-disable eslint-comments/no-unlimited-disable */
/* eslint-disable */`
        }
      },
      {
        [`add`]: {
          content: `import * as ${namespacedImportName} from '@shopify/hydrogen/storefront-api-types';
`
        }
      },
      {
        [`typescript-operations`]: {
          skipTypename: true,
          useTypeImports: true,
          preResolveTypes: false
        }
      },
      { [`gen-dts`]: { sourcesWithOperations, interfaceExtensionCode } },
      ...options.plugins
    ];
    return [
      {
        filename: options.baseOutputDir,
        plugins,
        pluginMap,
        schema: options.schema,
        config: {
          ...options.config,
          namespacedImportName
        },
        documents: sources,
        documentTransforms: options.documentTransforms
      }
    ];
  }
};

exports.interfaceExtensionCode = interfaceExtensionCode;
exports.namespacedImportName = namespacedImportName;
exports.preset = preset;
