// @shopify/hydrogen-codegen/dist/esm/preset.js
import * as addPlugin from '@graphql-codegen/add';
import * as typescriptOperationPlugin from '@graphql-codegen/typescript-operations';
import { processSources } from './sources.js';
import { plugin } from './plugin.js';

const namespacedImportName = "StorefrontAPI";
const interfaceExtensionCode = `
declare module '@shopify/hydrogen' {
  interface StorefrontQueries extends GeneratedQueryTypes {}
  interface StorefrontMutations extends GeneratedMutationTypes {}
}`;
const preset = {
  buildGeneratesSection: (options) => {
    if (!options.baseOutputDir.endsWith(".d.ts")) {
      throw new Error("[hydrogen-preset] target output should be a .d.ts file");
    }
    if (options.plugins?.length > 0 && Object.keys(options.plugins).some((p) => p.startsWith("typescript"))) {
      throw new Error(
        "[hydrogen-preset] providing additional typescript-based `plugins` leads to duplicated generated types"
      );
    }
    const sourcesWithOperations = processSources(options.documents);
    const sources = sourcesWithOperations.map(({ source }) => source);
    const pluginMap = {
      ...options.pluginMap,
      [`add`]: addPlugin,
      [`typescript-operations`]: typescriptOperationPlugin,
      [`gen-dts`]: { plugin: plugin }
    };
    const plugins = [
      {
        [`add`]: {
          content: `/* eslint-disable eslint-comments/disable-enable-pair */
/* eslint-disable eslint-comments/no-unlimited-disable */
/* eslint-disable */`
        }
      },
      {
        [`add`]: {
          content: `import * as ${namespacedImportName} from '@shopify/hydrogen/storefront-api-types';
`
        }
      },
      {
        [`typescript-operations`]: {
          skipTypename: true,
          useTypeImports: true,
          preResolveTypes: false
        }
      },
      { [`gen-dts`]: { sourcesWithOperations, interfaceExtensionCode } },
      ...options.plugins
    ];
    return [
      {
        filename: options.baseOutputDir,
        plugins,
        pluginMap,
        schema: options.schema,
        config: {
          ...options.config,
          namespacedImportName
        },
        documents: sources,
        documentTransforms: options.documentTransforms
      }
    ];
  }
};

export { interfaceExtensionCode, namespacedImportName, preset };
