// @shopify/cli-hydrogen/dist/commands/hydrogen/shortcut.js
import Command from '@shopify/cli-kit/node/base-command';
import { renderSuccess, renderFatalError } from '@shopify/cli-kit/node/ui';
import { isWindows, isGitBash, supportsShell, hasAlias, shellWriteFile, homeFileExists, shellRunScript } from '../../lib/shell.js';

const ALIAS_NAME = "h2";
class Shortcut extends Command {
  static description = `Creates a global \`${ALIAS_NAME}\` shortcut for the Hydrogen CLI`;
  async run() {
    await runCreateShortcut();
  }
}
async function runCreateShortcut() {
  const shortcuts = isWindows() && !isGitBash() ? await createShortcutsForWindows() : await createShortcutsForUnix();
  if (shortcuts.length > 0) {
    renderSuccess({
      headline: `Shortcut ready for the following shells: ${shortcuts.join(
        ", "
      )}.
Restart your terminal session and run \`${ALIAS_NAME}\` from your local project.`
    });
  } else {
    renderFatalError({
      name: "error",
      type: 0,
      message: "No supported shell found.",
      tryMessage: "Please create a shortcut manually."
    });
  }
}
const BASH_ZSH_COMMAND = `
# Shopify Hydrogen alias to local projects
alias ${ALIAS_NAME}='$(npm prefix -s)/node_modules/.bin/shopify hydrogen'`;
const FISH_FUNCTION = `
function ${ALIAS_NAME} --wraps='shopify hydrogen' --description 'Shortcut for the Hydrogen CLI'
   set npmPrefix (npm prefix -s)
   $npmPrefix/node_modules/.bin/shopify hydrogen $argv
end
`;
async function createShortcutsForUnix() {
  const shells = [];
  if (supportsShell("zsh") && (hasAlias(ALIAS_NAME, "~/.zshrc") || shellWriteFile("~/.zshrc", BASH_ZSH_COMMAND, true))) {
    shells.push("zsh");
  }
  if (supportsShell("bash") && (hasAlias(ALIAS_NAME, "~/.bashrc") || shellWriteFile("~/.bashrc", BASH_ZSH_COMMAND, true))) {
    shells.push("bash");
  }
  if (supportsShell("fish") && await homeFileExists("~/.config/fish/functions") && shellWriteFile(`~/.config/fish/functions/${ALIAS_NAME}.fish`, FISH_FUNCTION)) {
    shells.push("fish");
  }
  return shells;
}
const PS_FUNCTION = `function Invoke-Local-H2 {$npmPrefix = npm prefix -s; Invoke-Expression "$npmPrefix\\node_modules\\.bin\\shopify.ps1 hydrogen $Args"}; Set-Alias -Name ${ALIAS_NAME} -Value Invoke-Local-H2`;
const PS_APPEND_PROFILE_COMMAND = `
if (!(Test-Path -Path $PROFILE)) {
  New-Item -ItemType File -Path $PROFILE -Force
}

$profileContent = Get-Content -Path $PROFILE
if (!$profileContent -or $profileContent -NotLike '*Invoke-Local-H2*') {
  Add-Content -Path $PROFILE -Value '${PS_FUNCTION}'
}
`;
async function createShortcutsForWindows() {
  const shells = [];
  if (shellRunScript(PS_APPEND_PROFILE_COMMAND, "powershell.exe")) {
    shells.push("PowerShell");
  }
  if (shellRunScript(PS_APPEND_PROFILE_COMMAND, "pwsh.exe")) {
    shells.push("PowerShell 7+");
  }
  return shells;
}

export { Shortcut as default, runCreateShortcut };
